src/core.jl

```diff
@@ -1793,7 +1793,7 @@ function set_tensor_shape(tensor::AbstractTensor, dims)
 end
 
 function get_tensor_num_dims(tensor::AbstractTensor)
-    if tensor.op.op_name == "NoOp"
+    if tensor isa Tensor && tensor.op.op_name == "NoOp"
```

Comment:
How about just adding a new method to `get_tensor_num_dims` instead of the `isa` check?

Reply:
Could do, but other than this bit it would be the same.

The only reason to do the `isa` is that `Variable`s do no have a `.op` field so we can't check if they are `"NoOP"`.
This whole condition will compile away during specialisation.